Feat/add gecko functions

BenjaSanchez opened the pull request:
closes https://github.com/DD-DeCaF/scrum/issues/1085

Proteomics can now be added to the model as constraints using a `driven` card. To enable feasible simulations, a preliminar iterative flexibilization step is added that will make the model grow at least at a `minimal_growth` = inputted measurement - uncertainty. If growth rate is not provided, this step is skipped.

@kvikshaug @carrascomj I decided to skip for now the option of including `uptake_secretion_rates` as I realized that `apply_measurements` needs to be modified to properly match exchange reactions in ecModels (will create a separate card for this)

kvikshaug commented on src/simulations/modeling/driven.py:
```diff
@@ -165,3 +160,144 @@ def adjust_fluxes2model(
+    """
+    for protein_id, measure in measurements.items():
+        try:
+            rxn = model.reactions.get_by_id("prot_{}_exchange".format(protein_id))
```
Prefer f-strings over `.format` for consistency

BenjaSanchez replied:
done in commit [`1fe6f9a`](https://github.com/DD-DeCaF/simulations/pull/147/commits/1fe6f9ad49497691b7f7feea0d7a10233907d456)

kvikshaug commented on src/simulations/modeling/adapter.py:
```diff
@@ -487,6 +505,27 @@ def bounds(measurement, uncertainty):
+    for measure in proteomics:
+        try:
+            reaction = model.reactions.get_by_id(
+                "prot_{}_exchange".format(measure["identifier"])
```
Prefer f-strings over `.format` for consistency

BenjaSanchez replied:
done in commit [`1fe6f9a`](https://github.com/DD-DeCaF/simulations/pull/147/commits/1fe6f9ad49497691b7f7feea0d7a10233907d456)

kvikshaug commented on src/simulations/modeling/driven.py:
```diff
@@ -165,3 +160,144 @@ def adjust_fluxes2model(
+                new_row = pd.DataFrame(
+                    data={"met_id": met.id, "value": ub}, index=[protein_id]
+                )
+                prot_df = prot_df.append(new_row)
```
Alternative suggestion using cobrapy's `query` (it might make some optimizations, I'm not sure though):

```python
for met in model.metabolites.query(lambda m: protein_id in m.id):
    new_row = pd.DataFrame(
        data={"met_id": met.id, "value": ub}, index=[protein_id]
    )
    prot_df = prot_df.append(new_row)
```

BenjaSanchez replied:
done in commit [`1fe6f9a`](https://github.com/DD-DeCaF/simulations/pull/147/commits/1fe6f9ad49497691b7f7feea0d7a10233907d456) (doesn't seem to decrease performance)

kvikshaug commented on src/simulations/modeling/driven.py:
```diff
@@ -165,3 +160,144 @@ def adjust_fluxes2model(
+            (
+                index
+                for (index, d) in enumerate(proteomics)
+                if d["identifier"] == protein
```
What is `d` short for? It's not that hard to see where the variable comes from, but in general I'd recommend avoiding abbreviations to improve readability.

BenjaSanchez replied:
done in commit [`1fe6f9a`](https://github.com/DD-DeCaF/simulations/pull/147/commits/1fe6f9ad49497691b7f7feea0d7a10233907d456): hopefully `dic` is descriptive enough?

kvikshaug commented on src/simulations/modeling/driven.py:
```diff
@@ -165,3 +160,144 @@ def adjust_fluxes2model(
+            ),
+            None,
+        )
+        proteomics.pop(index)
```
`del proteomics[index]` is a bit more explicit when you don't care about returning the item.

BenjaSanchez replied:
done in commit [`1fe6f9a`](https://github.com/DD-DeCaF/simulations/pull/147/commits/1fe6f9ad49497691b7f7feea0d7a10233907d456)

kvikshaug commented on src/simulations/modeling/adapter.py:
```diff
@@ -487,6 +505,27 @@ def bounds(measurement, uncertainty):
                 }
             )
 
+    for measure in proteomics:
```
We should probably verify that we have an ecModel here too, like in the flexibilization step above (I would only add the warning here; since the above logic is dependent on whether growth rate was applied).

BenjaSanchez replied:
Fully agree, the warning makes more sense here. Done in commit [`51822fb`](https://github.com/DD-DeCaF/simulations/pull/147/commits/51822fb586bf93fdad06494f7d8387a193855aeb)

carrascomj commented on src/simulations/schemas.py:
```diff
@@ -54,6 +54,12 @@ class Metabolomics(Schema):
+    identifier = fields.String(required=True)  # UniProt code
+    measurement = fields.Float(required=True)  # mmol/gDw
+    uncertainty = fields.Float(required=True)  # mmol/gDw
+
```
Would it make sense to add a namespace here too, like in Metabolomics and Fluxomics?

BenjaSanchez replied:
@carrascomj as we are not using that field in any of the `adapter` or `driven` functions I would avoid it for keeping the code briefer. If we ever need it then of course we can introduce it.

BenjaSanchez commented:
@kvikshaug addressed all of your comments, thanks for the review! I also added some fixes of things I discovered while including your corrections (handle no `uncertainty` in proteomics, proper order of inputs in `apply_measurements`, save unmatched proteins as warnings and not errors).

Note: This branch should not be merged until the `identifier` field of proteomics matches in backend and frontend (see https://github.com/DD-DeCaF/scrum/issues/1167), otherwise we cannot do proper testing in frontend. Will add backend tests next week :)

BenjaSanchez commented:
@kvikshaug done with backend tests! expanded `test_measurements_adapter` to test proteomics on both a normal model (no change) and an ecModel. I used a fake proteomics dataset that forces the ecModel to flexibilize 2 proteins, and leaves 1 protein out that is not matched. With this all routines in `driven` are also tested, without introducing too much additional computational time (from 32 seconds to 45 seconds)

kvikshaug commented on tests/unit/test_adapter.py:
```diff
@@ -105,8 +105,60 @@ def test_measurements_adapter(iJO1366):
+    # 2 operations (1 proteomics + growth rate) + 1 warning (1 skipped protein) are
+    # expected:
+    assert len(operations) == 2
+    assert len(warnings) == 1
```
To keep the tests as small and encapsulated as possible, it could make sense to split it up and move this part to a separate test.

BenjaSanchez replied:
I wrote it like this bc in that section we are overall testing the same function (`test_measurements_adapter`), but if you think it's preferable to split will do :)

BenjaSanchez replied:
update: fixed in [`074f6f6`](https://github.com/DD-DeCaF/simulations/pull/147/commits/074f6f641120df7e6aecba44a4fcb2c4ddafae72)
